addons/binding/org.openhab.binding.facerecognition/src/main/java/org/openhab/binding/facerecognition/handler/FaceRecognitionHandler.java

```diff
@@ -0,0 +1,310 @@
+            job.cancel(true);
+        }
+
+        try {
+            httpClient.stop();
+
+        } catch (Exception e) {
+            logger.error("{}", e.getMessage());
+        }
+
+        faceCascade.close();
+    }
+
+    private void snapshot(final URI uri) {
+
+        try {
+            // File not found?
+            if (faceCascade.empty()) {
+                throw new FileNotFoundException("\"" + FACE_CLASSIFIER_FILE + "\" not found!");
+            }
+
+            // Try to get image from camera
+            Request request = httpClient.newRequest(uri);
+            request.accept("image/jpeg");
+            ContentResponse response = request.send();
+
+            // Error occurred?
+            if (response.getStatus() != HttpStatus.OK_200) {
+                throw new HttpResponseException(response.getReason(), response);
+            }
+
+            // Get image from HTTP response
+            byte[] imageData = response.getContent();
+
+            // Publish raw image?
+            if (isLinked(CHANNEL_IMAGE)) {
+                updateState(imageChannelUID, new RawType(imageData, "image/jpeg"));
+            }
+
+            // Convert image to gray-scale
+            Mat imageDataMat = new Mat(imageData);
+            Mat grayScaleImage = imdecode(imageDataMat, CV_LOAD_IMAGE_GRAYSCALE);
+
+            // Detect face in image
+            RectVector faces = new RectVector();
+            faceCascade.detectMultiScale(grayScaleImage, faces);
+
+            // Try to find known face in image
+            for (int i = 0; i < faces.size(); i++) {
+
+                // Extract face from image
+                Rect faceRect = faces.get(i);
+                Mat face = new Mat(grayScaleImage, faceRect);
+
+                IntPointer label = new IntPointer(1);
+                DoublePointer score = new DoublePointer(1);
+
+                File file = new File(MODELS_FILE);
+                boolean facesTrained = file.exists();
+
+                // Faces trained yet?
+                if (facesTrained) {
+
+                    // Try to recognize user
+                    faceRecognizer.predict(face, label, score);
+                } else {
+                    logger.debug("No faces trained yet");
+
+                    // Set default unknown user values
+                    label.put(UNKNOWN_USER_LABEL);
+                    score.put(Double.MAX_VALUE);
+                }
+
+                // User recognized?
+                if (facesTrained && (score.get(0) <= threshold.intValue())) {
+
+                    if (isLinked(CHANNEL_ACCESS_GRANTED_IMAGE)) {
+                        ByteBuffer outputBuffer = ByteBuffer.allocate(face.arraySize());
+                        imencode(".jpg", face, outputBuffer);
+                        updateState(accessGrantedImageChannelUID, new RawType(outputBuffer.array(), "image/jpeg"));
+                    }
+
+                    if (isLinked(CHANNEL_ACCESS_GRANTED_USERNAME)) {
```

Comment:
You really don't need to check for a channel if it is linked. Just update it. Only if it costs a lot of resources to compute a channel (like with the `CHANNEL_ACCESS_GRANTED_IMAGE` one) you might check for a linkage.